File: v2/src/main/mongo/namespaces.ts
import type { Document, MongoClient } from 'mongodb';
import type {
  CollStats,
  CollectionInfo,
  CreateCollectionInput,
  DatabaseInfo,
  DbStats,
} from '../../shared/namespace.js';

export async function listDatabases(client: MongoClient): Promise<DatabaseInfo[]> {
  const result = (await client.db('admin').command({ listDatabases: 1 })) as Document;
  const dbs = (result['databases'] as Document[] | undefined) ?? [];
  return dbs.map((d) => ({
    name: String(d['name']),
    sizeOnDisk: typeof d['sizeOnDisk'] === 'number' ? d['sizeOnDisk'] : undefined,
    empty: typeof d['empty'] === 'boolean' ? d['empty'] : undefined,
  }));
}

export async function listCollections(
  client: MongoClient,
  database: string,
): Promise<CollectionInfo[]> {
  const raw = await client
    .db(database)
    .listCollections({}, { nameOnly: false })
    .toArray();
  return raw.map((c) => {
    const opts = ((c as Document)['options'] ?? {}) as Record<string, unknown>;
    const type: CollectionInfo['type'] =
      c.type === 'view'
        ? 'view'
        : opts['timeseries']
          ? 'timeseries'
          : 'collection';
    return { name: c.name, type, options: opts };
  });
}

export async function dbStats(client: MongoClient, database: string): Promise<DbStats> {
  const stats = (await client.db(database).command({ dbStats: 1, scale: 1 })) as Document;
  return {
    db: database,
    collections: Number(stats['collections'] ?? 0),
    views: Number(stats['views'] ?? 0),
    objects: Number(stats['objects'] ?? 0),
    avgObjSize: Number(stats['avgObjSize'] ?? 0),
    dataSize: Number(stats['dataSize'] ?? 0),
    storageSize: Number(stats['storageSize'] ?? 0),
    indexes: Number(stats['indexes'] ?? 0),
    indexSize: Number(stats['indexSize'] ?? 0),
    totalSize: Number(stats['totalSize'] ?? 0),
  };
}

export async function collStats(
  client: MongoClient,
  database: string,
  collection: string,
): Promise<CollStats> {
  const stats = (await client
    .db(database)
    .command({ collStats: collection, scale: 1 })) as Document;
  return {
    ns: `${database}.${collection}`,
    count: Number(stats['count'] ?? 0),
    size: Number(stats['size'] ?? 0),
    avgObjSize: Number(stats['avgObjSize'] ?? 0),
    storageSize: Number(stats['storageSize'] ?? 0),
    totalIndexSize: Number(stats['totalIndexSize'] ?? 0),
    nindexes: Number(stats['nindexes'] ?? 0),
    capped: Boolean(stats['capped']),
  };
}

export async function createDatabase(
  client: MongoClient,
  database: string,
): Promise<void> {
  // Mongo creates the db lazily on first write. We create a placeholder
  // collection and immediately drop it so the database becomes visible.
  await client.db(database).createCollection('_mex_init');
  await client.db(database).collection('_mex_init').drop();
}

export async function dropDatabase(client: MongoClient, database: string): Promise<void> {
  await client.db(database).dropDatabase();
}

export async function createCollection(
  client: MongoClient,
  database: string,
  input: CreateCollectionInput,
): Promise<void> {
  const options: Record<string, unknown> = {};
  if (input.capped) {
    options['capped'] = true;
    options['size'] = input.capped.sizeBytes;
    if (input.capped.maxDocs) options['max'] = input.capped.maxDocs;
  }
  if (input.timeseries) {
    options['timeseries'] = input.timeseries;
  }
  if (input.validator) {
    options['validator'] = input.validator;
    if (input.validationLevel) options['validationLevel'] = input.validationLevel;
    if (input.validationAction) options['validationAction'] = input.validationAction;
  }
  await client.db(database).createCollection(input.name, options);
}

export async function dropCollection(
  client: MongoClient,
  database: string,
  collection: string,
): Promise<void> {
  await client.db(database).collection(collection).drop();
}

export async function renameCollection(
  client: MongoClient,
  database: string,
  from: string,
  to: string,
): Promise<void> {
  await client.db(database).renameCollection(from, to);
}

File: v3/src/main/kotlin/io/mex/mongo/Namespaces.kt
package io.mex.mongo

import com.mongodb.client.MongoClient
import org.bson.Document

data class DatabaseInfo(val name: String, val sizeOnDisk: Long?, val empty: Boolean?)

enum class CollectionType { collection, view, timeseries }

data class CollectionInfo(
    val name: String,
    val type: CollectionType,
    val options: Map<String, Any?>,
)

data class DbStats(
    val db: String,
    val collections: Long,
    val views: Long,
    val objects: Long,
    val avgObjSize: Double,
    val dataSize: Long,
    val storageSize: Long,
    val indexes: Long,
    val indexSize: Long,
    val totalSize: Long,
)

data class CollStats(
    val ns: String,
    val count: Long,
    val size: Long,
    val avgObjSize: Double,
    val storageSize: Long,
    val totalIndexSize: Long,
    val nindexes: Long,
    val capped: Boolean,
)

data class CreateCollectionInput(
    val name: String,
    val cappedSize: Long? = null,
    val cappedMax: Long? = null,
    val timeseriesField: String? = null,
)

fun listDatabases(client: MongoClient): List<DatabaseInfo> {
    val result = client.getDatabase("admin").runCommand(Document("listDatabases", 1))
    @Suppress("UNCHECKED_CAST")
    val dbs = (result["databases"] as? List<Document>).orEmpty()
    return dbs.map {
        DatabaseInfo(
            name = it.getString("name"),
            sizeOnDisk = it.get("sizeOnDisk") as? Long,
            empty = it["empty"] as? Boolean,
        )
    }
}

fun listCollections(client: MongoClient, db: String): List<CollectionInfo> {
    val list = client.getDatabase(db).listCollections().toList()
    return list.map { doc ->
        val opts = (doc["options"] as? Document)?.toMap().orEmpty()
        val type = when {
            doc.getString("type") == "view" -> CollectionType.view
            opts.containsKey("timeseries") -> CollectionType.timeseries
            else -> CollectionType.collection
        }
        CollectionInfo(name = doc.getString("name"), type = type, options = opts)
    }
}

fun dbStats(client: MongoClient, db: String): DbStats {
    val s = client.getDatabase(db).runCommand(Document("dbStats", 1).append("scale", 1))
    return DbStats(
        db = db,
        collections = s.numberOr("collections"),
        views = s.numberOr("views"),
        objects = s.numberOr("objects"),
        avgObjSize = (s["avgObjSize"] as? Number)?.toDouble() ?: 0.0,
        dataSize = s.numberOr("dataSize"),
        storageSize = s.numberOr("storageSize"),
        indexes = s.numberOr("indexes"),
        indexSize = s.numberOr("indexSize"),
        totalSize = s.numberOr("totalSize"),
    )
}

fun collStats(client: MongoClient, db: String, collection: String): CollStats {
    val s = client.getDatabase(db)
        .runCommand(Document("collStats", collection).append("scale", 1))
    return CollStats(
        ns = "$db.$collection",
        count = s.numberOr("count"),
        size = s.numberOr("size"),
        avgObjSize = (s["avgObjSize"] as? Number)?.toDouble() ?: 0.0,
        storageSize = s.numberOr("storageSize"),
        totalIndexSize = s.numberOr("totalIndexSize"),
        nindexes = s.numberOr("nindexes"),
        capped = (s["capped"] as? Boolean) ?: false,
    )
}

/**
 * MongoDB has no explicit create-database command and does not list empty
 * databases, so a database only becomes visible once it holds a collection.
 * The initial collection is therefore kept, not dropped.
 */
fun createDatabase(client: MongoClient, db: String, initialCollection: String) {
    client.getDatabase(db).createCollection(initialCollection)
}

fun dropDatabase(client: MongoClient, db: String) {
    client.getDatabase(db).drop()
}

fun createCollection(client: MongoClient, db: String, input: CreateCollectionInput) {
    val options = com.mongodb.client.model.CreateCollectionOptions()
    if (input.cappedSize != null) {
        options.capped(true).sizeInBytes(input.cappedSize)
        if (input.cappedMax != null) options.maxDocuments(input.cappedMax)
    }
    if (input.timeseriesField != null) {
        options.timeSeriesOptions(
            com.mongodb.client.model.TimeSeriesOptions(input.timeseriesField),
        )
    }
    client.getDatabase(db).createCollection(input.name, options)
}

fun dropCollection(client: MongoClient, db: String, name: String) {
    client.getDatabase(db).getCollection(name).drop()
}

fun renameCollection(client: MongoClient, db: String, from: String, to: String) {
    val ns = com.mongodb.MongoNamespace(db, to)
    client.getDatabase(db).getCollection(from).renameCollection(ns)
}

private fun Document.numberOr(key: String): Long = (this[key] as? Number)?.toLong() ?: 0L
